server/src/main/scala/org/http4s/server/ServerApp.scala

```diff
@@ -0,0 +1,96 @@
+package org.http4s
+package server
+
+import java.net.ServerSocket
+import java.util.concurrent.atomic.AtomicReference
+import java.util.concurrent.{ CountDownLatch, RejectedExecutionException }
+import java.util.concurrent.atomic.AtomicBoolean
+import java.util.concurrent.{ Executors, ThreadFactory }
+import scala.annotation.tailrec
+import scalaz.concurrent.Task
+import org.http4s.util.threads
+
+/**
+ * Apps extending the server app trait get a graceful shutdown.  The
+ * 
+ */ 
+trait ServerApp {
+  private[this] val logger = org.log4s.getLogger
+
+  /** Return a server to run */
+  def server(args: List[String]): Task[Server]
```

Comment:
Yeah, I figured this was our chance to get away from a Java mistake.  I took inspiration from Scalaz's `TaskApp`, except I left an abstract method instead of letting the implementor pick one (and hopefully only one) override.